### PR TITLE
improve the loading component

### LINKED ISSUE #6071: [Accessibility]A11y__AzureDataStudio_CreateExternalTable_ScreenReader:Voice Over is not announcing status while page is loading state.
[Check out Keros!](https://microsoft.sharepoint.com/teams/msenable/Pages/accessibility-tools.aspx) - Identify accessibility bugs before check-in and make bug fixing faster and easier.

GitHubTags: #A11yMAS; #A11yTCS; #MAS1.3.1; #A11y_AzureDataStudio; #Win32; #DesktopApp

**Environment Details:**
Application name: Azure Data Studio
Version: 1.9.0-insider (user setup)
Commit: 2ca39b5
Electron: 3.1.8
Node.js: 10.2.0
V8: 6.6.346.32
OS: MacOS Mojave 10.14.5
**Additional Details:**
Tools Name: Voice Over
MAS Number :MAS1.3.1
**Steps to Reproduce** :
1.Install the app Azure Data Studio.
2.Enable SQL server2019 preview->select table from the database.
3.Start Voice Over
4.Observe that while page is loading state ,screen reader doesn't announce "Page is still loading".
**Actual** :
When page is in loading state ,screen reader is not announcing "Page is still loading ".It only announces the information which present on that page.
**Expected**:
Screen reader should announce "Page is still loading " when page is loading.
**Recommendation**:
Please refer the following snippet link for fixes:
https://microsoft.sharepoint.com/teams/msenable/mas/pages/browse-fixes.aspx
**User Impact**:
Screen reader only user won't understand that page is loading and may misunderstood that page does not contain any content.
**MAS Reference**:
**MAS1.3.1**-
https://microsoft.sharepoint.com/:w:/r/teams/msenable/_layouts/15/WopiFrame.aspx?sourcedoc={54f28d1f-a2d1-4dcd-84e1-5c9b87e8aba4}
**Attachment for Reference**:
[6071_[Accessibility]A11y__AzureDataStudio_CreateExternalTable_ScreenReader.pptx](https://github.com/microsoft/azuredatastudio/files/3300921/6071_.Accessibility.A11y__AzureDataStudio_CreateExternalTable_ScreenReader.pptx)
Does this issue occur when all extensions are disabled?: Yes

### PR DESCRIPTION
**Pull request by alanrenmsft**
This PR fixes #6071 
@Benjin previously made a change to loading component so that the loading message can be announced, but the loading completed is not announced, plus, after reading the description of alert role and the status role, i think status role properly is a better choice. with this change, screen reader will read both loading and loading completed. feature owner can also customize the loading and loading completed texts.

also previously i received comments during scenario reviews suggesting to also add text next to the spinner so that user knows what is being loaded, this PR also addresses it. by default the text will not be displayed.

Screen Shot 2019-11-06 at 4 37 22 PM

### CODE REVIEW

File: src/sql/workbench/browser/modelComponents/loadingComponent.component.ts [hunk 1]
```diff
@@ -12,20 +12,23 @@ import * as azdata from 'azdata';
...
 import { IComponent, IComponentDescriptor, IModelStore } from 'sql/workbench/browser/modelComponents/interfaces';
-import * as nls from 'vs/nls';
+import { localize } from 'vs/nls';
 
 @Component({
 	selector: 'modelview-loadingComponent',
 	template: `
-		<div class="modelview-loadingComponent-container" role="alert" aria-busy="true" *ngIf="loading">
-			<div class="modelview-loadingComponent-spinner" *ngIf="loading" [title]=_loadingTitle #spinnerElement></div>
+		<div role="status" aria-live="polite" class="modelview-loading-component-status-message">
```
Comment (Benjin) [inline src/sql/workbench/browser/modelComponents/loadingComponent.component.ts]: No alert role?  I found that setting aria-live wasn't enough unless the contents of the div themselves were changing after the components were loaded.

File: src/sql/azdata.d.ts [hunk 2]
```diff
@@ -3238,6 +3238,9 @@ declare module 'azdata' {
 
 	export interface LoadingComponentProperties {
 		loading?: boolean;
+		showText?: boolean;
```
[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply src/sql/azdata.d.ts]: Eh, thinking about it I'm not sure that's actually something we should care about changing after initial creation - it definitely seems weird. Maybe we should start thinking about having some read-only properties.

Anyways I'm fine either way so do whatever you think is better.

File: src/sql/workbench/browser/modelComponents/loadingComponent.component.ts (hunk 1 above)
Comment (alanrenmsft) [reply src/sql/workbench/browser/modelComponents/loadingComponent.component.ts]: I am using status role now, it is also a live region role similar to alert. https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_status_role

File: src/sql/azdata.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (alanrenmsft) [reply src/sql/azdata.d.ts]: right, this should be something you set at creation time, I couldn't think of a scenario we need to update this after creation.

**Comment by alanrenmsft** [comment]
> It was definitely working when I evaluated the fix; wonder whether it was an incomplete fix, or something regressed.

your fix worked for me, the screen reader will read the loading message, but we are not sending the loading completed message. this is an improvement to what you have done. :)

**Comment by alanrenmsft** [comment]
> So how is this announced to the user? When they first enter the page is anything announced? Or does it only happen when they select the loading component in reading mode? (the loading spinner is still skipped for tab selection right?)

it will read loading (or the custom loadingText) and once the loading is done it will read loading completed (or the custom loading completed text)

File: extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts [hunk 3]
```diff
@@ -97,7 +97,11 @@ export class ResourceTypePickerDialog extends DialogBase {
```
Comment (alanrenmsft) [inline extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts]: i didn't mean to include changes to this file, will revert the changes

**Comment by Charles-Gagnon** [comment]
@alanrenmsft Yeah - I was asking if it'll automatically read the text when you enter the page or only when you select the item in reading mode. What happens with multiple loading components on the same page if the former?

**Comment by alanrenmsft** [comment]
> @alanrenmsft Yeah - I was asking if it'll automatically read the text when you enter the page or only when you select the item in reading mode. What happens with multiple loading components on the same page if the former?

it will be announced by screen reader without needing to be focused. if there are multiple, then you will hear multiple loading/loading completed. that is once of the reason i added the custom loading/loading completed text so that it is clear what the content is about.

**Comment by Benjin** [comment]
@alanrenmsft

>we [were] not sending the loading completed message

Ah, true.  I think the way the previous fix worked was that it would always announce "Loading", but each loaded content pane needed a code change to announce the new stuff.
